cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -92,45 +94,146 @@ func (c *machineController) findMachineSet(machine *v1alpha1.Machine) (*v1alpha1
 		return nil, fmt.Errorf("internal error; unexpected type: %T", machineSet)
 	}
 
+	if !machineBelongsToMachineSet(machine, machineSet) {
+		return nil, fmt.Errorf("internal error; owner references do not match")
+	}
+
 	return machineSet.DeepCopy(), nil
 }
 
-// run starts shared informers and waits for the shared informer cache
-// to synchronize.
+// run starts shared informers and waits for the informer cache to
+// synchronize.
 func (c *machineController) run(stopCh <-chan struct{}) error {
-	c.informerFactory.Start(stopCh)
+	c.kubeInformerFactory.Start(stopCh)
+	c.clusterInformerFactory.Start(stopCh)
+
+	glog.V(4).Infof("waiting for caches to sync")
+	if !cache.WaitForCacheSync(stopCh,
+		c.nodeInformer.HasSynced,
+		c.machineInformer.Informer().HasSynced,
+		c.machineSetInformer.Informer().HasSynced) {
+		return fmt.Errorf("syncing caches failed")
+	}
+
+	return nil
+}
 
-	glog.V(4).Infof("waiting for machine cache to sync")
-	if !cache.WaitForCacheSync(stopCh, c.machineInformer.Informer().HasSynced) {
-		return fmt.Errorf("cannot sync machine cache")
+// findMachineByNodeProviderID lookups a Node by node.Spec.ProviderID.
+// If the node object is found it looks for a "machine" annotation on
+// the found node to find the underlying machine object. Returns nil
+// if either the Node by node.Spec.ProviderID cannot be found or if
+// the node has no "machine" annotation. A DeepCopy() of the object is
+// returned on success.
+func (c *machineController) findMachineByNodeProviderID(node *apiv1.Node) (*v1alpha1.Machine, error) {
+	objs, err := c.nodeInformer.GetIndexer().ByIndex(nodeProviderIDIndex, node.Spec.ProviderID)
+	if err != nil {
+		return nil, err
+	}
+	if len(objs) != 1 {
+		return nil, nil
 	}
 
-	glog.V(4).Infof("waiting for machineset cache to sync")
-	if !cache.WaitForCacheSync(stopCh, c.machineSetInformer.Informer().HasSynced) {
-		return fmt.Errorf("cannot sync machineset cache")
+	node, ok := objs[0].(*apiv1.Node)
+	if !ok {
+		return nil, fmt.Errorf("internal error; unexpected type %T", node)
 	}
 
-	return nil
+	machineName, found := node.Annotations["machine"]
```

Comment:
Seems like this should probably try the newer `cluster.k8s.io/machine` annotation first, then fallback to `machine`.

Reply:
Done. Thanks for the link.